File: Data Structures & Algorithms/depth-of-binary-tree/submission-3.kt
/**
 * Definition for a binary tree node.
 * class TreeNode(var `val`: Int) {
 *     var left: TreeNode? = null
 *     var right: TreeNode? = null
 * }
 */

class Solution {
    fun maxDepth(root: TreeNode?, depth: Int = 0): Int {
        if(root == null) {
            return depth
        }

        val left = maxDepth(root.left, depth + 1)
        var right = maxDepth(root.right, depth + 1)

        return Math.max(left, right)
    }
}

File: Data Structures & Algorithms/count-good-nodes-in-binary-tree/submission-3.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right

class Solution:
    def goodNodes(self, root: TreeNode) -> int:
        def dfs(node: Optional['TreeNode'], max_val: float):
            if not node:
                return 0

            count = 0

            if node.val >= max_val:
                count += 1

            count += dfs(node.left, max(max_val, node.val))
            count += dfs(node.right, max(max_val, node.val))

            return count

        return dfs(root, root.val)